In [1]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: AuthenticAMD
cpu family	: 23
model		: 49
model name	: AMD EPYC 7B12
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2249.998
cache size	: 512 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr arat npt nrip_save umip rdpid
bugs		: sysret_ss_attrs null_seg spectre_v1 spectre_v2 spec_store_bypass retbleed smt_rsb
bogomips	: 4499.

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [2]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [3]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

2 3
5

### Test input output files

In [4]:
%%writefile input1.txt
4 8

Writing input1.txt


In [5]:
%%writefile output1.txt
12

Writing output1.txt


In [6]:
%%writefile input2.txt
0 9

Writing input2.txt


In [7]:
%%writefile output2.txt
9

Writing output2.txt


In [14]:
!./APlusB < input1.txt

12

In [8]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [15]:
!grep -xvFf output1.txt output1.txt

In [16]:
!grep -xvFf output1.txt output2.txt

9


In [17]:
!grep -xvFf output2.txt output1.txt

12


In [18]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [19]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [132]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [133]:
!python sum.py

4 8
12


In [134]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

Traceback (most recent call last):
  File "/content/sum.py", line 7, in <module>
    a, b = map(int, input().split())
ValueError: not enough values to unpack (expected 2, got 1)


In [23]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [24]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [25]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

2
1
2
2


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [26]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [27]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [28]:
%%writefile output1.txt
90

Overwriting output1.txt


In [29]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [30]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [31]:
%%writefile output2.txt
90

Overwriting output2.txt


In [32]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [33]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [34]:
200000*200000

40000000000

In [35]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [36]:
!./max_pairwise_product_full < input3.txt

1345294336


In [37]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [38]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [39]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [40]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [41]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [42]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [43]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [45]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [47]:
%%writefile input4.txt
7
1 3 200000 9 200000 23 200000


Writing input4.txt


In [48]:
%%writefile output4.txt
40000000000

Writing output4.txt


In [49]:
%%writefile input5.txt
2
1 1

Writing input5.txt


In [50]:
%%writefile output5.txt
1

Writing output5.txt


In [51]:
%%writefile input6.txt
8
2 2 2 2 2 2 2 2

Writing input6.txt


In [52]:
%%writefile output6.txt
4

Writing output6.txt


In [53]:
%%writefile input7.txt
5
1 3 200000 9 200000

Writing input7.txt


In [54]:
%%writefile output7.txt
40000000000

Writing output7.txt


In [55]:
%%writefile input8.txt
3
9 8 9

Writing input8.txt


In [56]:
%%writefile output8.txt
81

Writing output8.txt


In [57]:
%%writefile input9.txt
4
200000 200000 200000 200000

Writing input9.txt


In [58]:
%%writefile output9.txt
40000000000

Writing output9.txt


In [59]:
%%writefile input10.txt
3
0 0 0

Writing input10.txt


In [60]:
%%writefile output10.txt
0

Writing output10.txt


In [61]:
 %%writefile run_all_test_1.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test_1.sh


In [62]:
!chmod u+x run_all_test_1.sh
!./run_all_test_1.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [63]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [64]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [65]:
199999 * 200000

39999800000

In [66]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [67]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [68]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [69]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [70]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	3m20.348s
user	3m19.424s
sys	0m0.081s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m4.108s
user	4m57.506s
sys	0m0.172s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

589



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n= 10

560



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


200001000021


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

306/200001000010

1.5299923499617505e-09

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.873s
user	0m0.852s
sys	0m0.003s


## Python

In [87]:
%%writefile max_pairwise_product_long_full.py
def MaxPairwiseProduct(numbers):
  max_product = 0
  aux_product = 0
  n = len(numbers)

  for first in range (n):
    for second in range(n):
      if(first != second):
        aux_product = numbers[first]*numbers[second]
        if aux_product > max_product:
          max_product = aux_product
  return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))

Overwriting max_pairwise_product_long_full.py


In [90]:
!time python3 max_pairwise_product_long_full.py < input12.txt

5
1 2 3 4 5
4
4
39999800000

real	56m19.444s
user	55m47.680s
sys	0m1.225s


In [93]:
%%writefile max_pairwise_product_long_full_execints.py
def MaxPairwiseProduct(numbers):
  global execints
  max_product = 0
  execints+=1
  aux_product = 0
  execints+=1
  n = len(numbers)
  execints+=1

  execints+=1
  for first in range (n):
    execints+=1

    execints+=1
    for second in range(n):
      execints+=1

      execints+=1
      if(first != second):
        execints+=1
        aux_product = numbers[first]*numbers[second]
        execints+=1
        if aux_product > max_product:
          execints+=1
          max_product = aux_product
          execints+=1
  execints+=1
  return max_product


execints = 0
n = int(input())
execints+=1
numbers = [int(i) for i in input().split()]
execints+= n
MaxPairwiseProduct(numbers)
print(execints)

Overwriting max_pairwise_product_long_full_execints.py


In [95]:
!python3 max_pairwise_product_long_full_execints.py < input12.txt

160000200008


In [96]:
3379/160000200008

2.1118723600539563e-08

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 2.11*10^{-8} s = 1s$$

$$ n \approx 3000$$

In [99]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input14.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <3001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [100]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [101]:
!time python3 max_pairwise_product_long_full.py < input14.txt

8997000

real	0m0.741s
user	0m0.723s
sys	0m0.006s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [102]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 10000 <<"\n";  // n
 for (int i=1; i <10001; i++){
     MyFile << 10001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [103]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [104]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [105]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


Writing max_pairwise_product_long.cpp


In [106]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input15.txt

99990000

real	0m0.235s
user	0m0.231s
sys	0m0.001s


In [112]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    for (int first = 0; first < n; ++first) {                                //1+2n
        execints++;

        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
            execints++;

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            execints++;

            execints++;
            if (max_product < aux_product)  {                                //(n-1)n/2
                execints++;
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
    execints++;                                                               // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    long long execints =0LL;

    int n;                                                //1
    execints++;

    std::cin >> n;                                        //1
    execints++;

    std::vector<int> numbers(n);                          //n
    execints++;
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
        execints++;
    }

    std::cout << MaxPairwiseProduct(numbers, execints) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    execints++;
    std::cout << execints << "\n";
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound

Overwriting max_pairwise_product_long_execints.cpp


In [113]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input15.txt

99990000
150005009


In [114]:
0.235 / 150005009

1.5666143521913991e-09

# Estimate the maximun n that can be solved in 1 second

$$(2.5n^2 + 7n) \times 1.56 *10^{-9} s = 1s$$

$$ n \approx 16000$$

In [118]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input16.txt");

 MyFile << 10000 <<"\n";  // n
 for (int i=1; i <16001; i++){
     MyFile << 16001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [119]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [120]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input16.txt

255984000

real	0m0.232s
user	0m0.226s
sys	0m0.003s


In [107]:
%%writefile max_pairwise_product_long.py
def MaxPairwiseProduct(numbers):
  max_product = 0
  aux_product = 0
  n = len(numbers)

  for first in range (n):
    for second in range(first+1, n):
      aux_product = numbers[first]*numbers[second]
      if aux_product > max_product:
        max_product = aux_product
  return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))


Writing max_pairwise_product_long.py


In [108]:
!time python3 max_pairwise_product_long.py < input15.txt

99990000

real	0m4.209s
user	0m4.152s
sys	0m0.005s


In [109]:
%%writefile max_pairwise_product_long_execints.py
def MaxPairwiseProduct(numbers):
  global execints
  max_product = 0
  execints+=1

  aux_product = 0
  execints+=1

  n = len(numbers)
  execints+=1

  execints+=1
  for first in range (n):

    execints+=1
    for second in range(first+1, n):

      aux_product = numbers[first]*numbers[second]
      execints+=1

      if aux_product > max_product:
        execints+=1
        max_product = aux_product
        execints+=1

  execints+=1
  return max_product

execints = 0
n = int(input())
execints+=1
numbers = [int(i) for i in input().split()]
execints+=2*n
print(execints)

Writing max_pairwise_product_long_execints.py


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [128]:
%%writefile max_pairwise_product_long_sort.cpp
#include <iostream>
#include <algorithm>
#include <vector>


int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //1
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];                           //n
    }
    std::sort(numbers.begin(), numbers.end());              // n lg n

    std::cout << numbers[n-1] * numbers[n-2] << "\n";     // 4 + n lg n + n
    return 0;                                             // --------------
}                                                       //


Overwriting max_pairwise_product_long_sort.cpp


Test with inputs 1 and 2



In [129]:
!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!./max_pairwise_product_long_sort < input1.txt

90


In [130]:
!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!./max_pairwise_product_long_sort < input2.txt

90


Test and time it with input 3

In [135]:
!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!time ./max_pairwise_product_long_sort < input3.txt

1345294336

real	0m0.003s
user	0m0.002s
sys	0m0.000s


20 stress tests

In [136]:
 %%writefile run_all_test_2.sh
#!/bin/bash
g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
for i in {4..10}
do
    ./max_pairwise_product_long_sort < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test_2.sh


In [138]:
!chmod u+x run_all_test_2.sh
!./run_all_test_2.sh

1345294336
1345294336
1345294336


Count executed instructions in code and verify with inputs 1,2 and 3


In [141]:
%%writefile max_pairwise_product_long_sort_execints.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>

int main() {                                              //N inst exec by line executed
    int execints = 0;
    int n;
    execints++;

    std::cin >> n;                                        //1
    execints++;

    std::vector<int> numbers(n);                          //1
    execints++;

    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];                           //n
        execints++;
    }

    std::sort(numbers.begin(), numbers.end());              // n lg n
    execints = execints + (n * log(n));

    std::cout << numbers[n-1] * numbers[n-2] << "\n";     // 4 + n lg n + n
    std::cout << execints << "\n";     // 4 + n lg n + n
    return 0;                                             // --------------
}

Overwriting max_pairwise_product_long_sort_execints.cpp


In [143]:
!g++ max_pairwise_product_long_sort_execints.cpp -o max_pairwise_product_long_sort_execints
!./max_pairwise_product_long_sort_execints < input3.txt

1345294336
19


Calculate estimated time by instruction and number of instructions by second

In [144]:
0.03/19

0.0015789473684210526

# Estimate the maximun n that can be solved in 1 second

$$(n + n log n) \times 0.001579 s = 1s$$

$$ n \approx 110000$$

In [149]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input17.txt");

 MyFile << 110000 <<"\n";  // n
 for (int i=1; i <110001; i++){
     MyFile << 110001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [150]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [151]:
!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!time ./max_pairwise_product_long_sort < input17.txt

-785011888

real	0m0.069s
user	0m0.064s
sys	0m0.000s


### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [152]:
%%writefile max_pairwise_product_sort.py
def MaxPairwiseProduct(numbers):
  n = len(numbers)
  numbers.sort()
  return numbers[n-1] * numbers[n-2]

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [153]:
!python3 max_pairwise_product_sort.py < input1.txt
!python3 max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [154]:
!time python3 max_pairwise_product_sort.py < input3.txt

40000000000

real	0m0.039s
user	0m0.029s
sys	0m0.007s


20 stress tests

In [155]:
 %%writefile run_all_test_2.sh
#!/bin/bash
for i in {4..10}
do
    python3 max_pairwise_product_sort.py < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test_2.sh


In [156]:
!chmod u+x run_all_test_2.sh
!./run_all_test_2.sh

Count executed instructions in code and verify with inputs 1,2 and 3


In [166]:
%%writefile max_pairwise_product_sort.py
import math

def MaxPairwiseProduct(numbers):
  global execints
  n = len(numbers)
  execints+=1
  numbers.sort() # nlgn -- n
  execints += n * math.log(n)
  execints+=1
  return numbers[n-1] * numbers[n-2]

execints = 0
n = int(input())
execints+=1
numbers = [int(i) for i in input().split()]
execints+=1
print(MaxPairwiseProduct(numbers))
print(execints)

Overwriting max_pairwise_product_sort.py


In [167]:
! python3 max_pairwise_product_sort.py < input3.txt

40000000000
14.75055681536833


Calculate estimated time by instruction and number of instructions by second

In [168]:
0.039/15

0.0026

# Estimate the maximun n that can be solved in 1 second

$$(n + n log n) \times 0.0026 s = 1s$$

$$ n \approx 72000$$

### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [169]:
%%writefile max_pairwise_product_long_efficient.cpp

#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n;
    int indice = 0;
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    std::cin >> n;
    std::vector<int> numbers(n);

    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
        if(numbers[i] > aux_product){
            aux_product = numbers[i];
            indice = i;
        }
    }

    max_product = aux_product;
    aux_product = 0;

    for (int i = 0; i < n; ++i) {
        if(numbers[i] > aux_product){
            if(indice != i){
                aux_product = numbers[i];
            }
        }
    }

    std::cout << aux_product* max_product << "\n";
    return 0;
}


Writing max_pairwise_product_long_efficient.cpp


Test with inputs 1 and 2




In [170]:
!g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient
!./max_pairwise_product_long_efficient < input1.txt
!g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient
!./max_pairwise_product_long_efficient < input2.txt

90
90


Test and time it with input 3

In [174]:
!g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient
!time ./max_pairwise_product_long_efficient < input3.txt

40000000000

real	0m0.002s
user	0m0.001s
sys	0m0.000s


20 stress tests

In [172]:
 %%writefile run_all_test_3.sh
#!/bin/bash
g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient
for i in {4..10}
do
    ./max_pairwise_product_long_efficient < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test_3.sh


In [173]:
!chmod u+x run_all_test_3.sh
!./run_all_test_2.sh

17.621371043387192
5.386294361119891
20.635532333438686
12.047189562170502
7.295836866004329
9.545177444479563
7.295836866004329


Count executed instructions in code and verify with inputs 1,2 and 3

In [175]:
%%writefile max_pairwise_product_long_efficient_execints.cpp

#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int execints = 0;
    int n;
    execints++;

    int indice = 0;
    execints++;

    long long max_product = 0LL;
    execints++;

	  long long aux_product = 0LL;
    execints++;

    std::cin >> n;
    execints++;

    std::vector<int> numbers(n);
    execints++;

    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
        execints++;

        execints++;
        if(numbers[i] > aux_product){
            aux_product = numbers[i];
            execints++;
            indice = i;
            execints++;
        }
    }


    max_product = aux_product;
    execints++;

    aux_product = 0;
    execints++;

    for (int i = 0; i < n; ++i) {
      execints++;

        execints++;
        if(numbers[i] > aux_product){

            execints++;
            if(indice != i){
              execints++;
                aux_product = numbers[i];
            }
        }
    }

    std::cout << execints << "\n";
    return 0;
}

Writing max_pairwise_product_long_efficient_execints.cpp


In [176]:
!g++ max_pairwise_product_long_efficient_execints.cpp -o max_pairwise_product_long_efficient_execints
!./max_pairwise_product_long_efficient_execints < input3.txt

47


Calculate estimated time by instruction and number of instructions by second

In [177]:
0.002/47

4.2553191489361704e-05

# Estimate the maximun n that can be solved in 1 second

$$(2n) \times 4.255 *10^{-5} s = 1s$$

$$ n \approx 11000$$

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [178]:
%%writefile max_pairwise_product_efficient.py
n = int(input())
max = 0
max_mayor = 0
indice_mayor = 0
numbers = [int(i) for i in input().split()]

for i in range (len(numbers)):
  if numbers[i]>max:
    max = numbers[i]
    indice_mayor = i

max_mayor = max
max = 0

for i in range (len(numbers)):
  if numbers[i]>max and i != indice_mayor:
    max = numbers[i]

print(max_mayor*max)

Writing max_pairwise_product_efficient.py


Test with inputs 1 and 2




In [184]:
!python3 max_pairwise_product_efficient.py < input1.txt
!python3 max_pairwise_product_efficient.py < input2.txt

90
90


Test and time it with input 3

In [185]:
!time python3 max_pairwise_product_efficient.py < input3.txt

40000000000

real	0m0.031s
user	0m0.025s
sys	0m0.005s


20 stress tests

In [180]:
 %%writefile run_all_test_3.sh
#!/bin/bash
for i in {4..10}
do
    python3 max_pairwise_product_efficient.py < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test_3.sh


In [181]:
!chmod u+x run_all_test_3.sh
!./run_all_test_2.sh

17.621371043387192
5.386294361119891
20.635532333438686
12.047189562170502
7.295836866004329
9.545177444479563
7.295836866004329


Count executed instructions in code and verify with inputs 1,2 and 3

In [182]:
%%writefile max_pairwise_product_efficient_execints.py
execints = 0
n = int(input())
execints += 1
max = 0
execints += 1
max_mayor = 0
execints += 1
indice_mayor = 0
execints += 1
numbers = [int(i) for i in input().split()]
execints += 2 * n

for i in range (len(numbers)):
  execints += 1

  execints += 1
  if numbers[i]>max:
    execints += 1
    max = numbers[i]
    execints += 1
    indice_mayor = i

execints += 1
max_mayor = max
max = 0
execints += 1

for i in range (len(numbers)):
  execints += 1

  execints += 1
  if numbers[i]>max and i != indice_mayor:
    max = numbers[i]
    execints += 1

print(execints)

Writing max_pairwise_product_efficient_execints.py


In [183]:
!python3 max_pairwise_product_efficient_execints.py < input3.txt

52


Calculate estimated time by instruction and number of instructions by second

In [186]:
0.0321 / 52

0.0006173076923076923

# Estimate the maximun n that can be solved in 1 second

$$(2n) \times 0.000617 s = 1s$$

$$ n \approx 8100$$